includes/amp-helper-functions.php

```diff
@@ -406,17 +406,54 @@ function amp_print_schemaorg_metadata() {
  *
  * @since 0.7.0
  */
-function amp_prepare_comment_post() {
-	if ( ! isset( $_GET['__amp_source_origin'] ) ) { // WPCS: CSRF ok. Beware of AMP_Theme_Support::purge_amp_query_vars().
+function amp_handle_xhr_request() {
+	global $pagenow;
+	if ( isset( $_GET['__amp_redirect'] ) ) { // WPCS: CSRF ok.
+		add_action( 'template_redirect', function() {
+			// grab post data.
+			$transint_name = wp_unslash( $_GET['__amp_redirect'] ); // WPCS: CSRF ok, input var ok.
+			$_POST         = get_transient( $transint_name );
+			delete_transient( $transint_name );
+		}, 0 );
+	}
+	if ( ! isset( $_GET['__amp_source_origin'] ) || ! isset( $pagenow ) ) { // WPCS: CSRF ok. Beware of AMP_Theme_Support::purge_amp_query_vars().
 		return;
 	}
 
-	// Add amp comment hooks.
-	add_filter( 'comment_post_redirect', function() {
-		// We don't need any data, so just send a success.
-		wp_send_json_success();
-	}, PHP_INT_MAX, 2 );
+	if ( 'wp-comments-post.php' === $pagenow ) {
+		// This only runs on wp-comments-post.php.
+		add_filter( 'comment_post_redirect', function() {
+			// We don't need any data, so just send a success.
+			wp_send_json_success();
+		}, PHP_INT_MAX, 2 );
+		amp_handle_xhr_headers_output();
+	} elseif ( isset( $_GET['_wp_amp_action_xhr_converted'] ) ) { // WPCS: CSRF ok.
+		// Add amp redirect hooks.
+		add_filter( 'wp_redirect', 'amp_intercept_post_request_redirect', PHP_INT_MAX, 2 );
+		add_action( 'template_redirect', function() {
+			// grab post data.
+			$transient_name = uniqid();
+			set_transient( $transient_name, wp_unslash( $_POST ), 60 ); // WPCS: CSRF ok, input var ok.
```

Comment:
This is clever, but I do fear it could lead to unexpected results like you said. So I think for now we should omit it and instead revisit it later if we find that it is a common problem.

Reply:
@westonruter Ye I thought so to. I was going to just remove it, but wanted to see if I could at least make your example work. It's better not being there for now.